Update changelog to reflect 2.3.0 (#662)
Automattic/VIP-Coding-Standards: CHANGELOG.md

## Changed
- [#581](https://github.com/Automattic/VIP-Coding-Standards/pull/581): AlwaysReturnInFilter: flag abstract methods for manual inspection.
- [#584](https://github.com/Automattic/VIP-Coding-Standards/pull/584): Performance: more selective sniffing for efficiency.
- [#586](https://github.com/Automattic/VIP-Coding-Standards/pull/586): IncludingNonPHPFile: various improvements.
- [#587](https://github.com/Automattic/VIP-Coding-Standards/pull/587): LowExpiryCacheTime: new warning added for manual inspection along with various improvements.
- [#589](https://github.com/Automattic/VIP-Coding-Standards/pull/589): WPQueryParams: inherit from abstract class and flag for 'exclude' parameter.
- [#592](https://github.com/Automattic/VIP-Coding-Standards/pull/592): DynamicCalls: various improvements.
- [#595](https://github.com/Automattic/VIP-Coding-Standards/pull/595): Underscorejs: various improvements.
- [#620](https://github.com/Automattic/VIP-Coding-Standards/pull/620): Ruleset: silence UnusedVariable from VariableAnalysis to reduce noise.

> - [#624](https://github.com/Automattic/VIP-Coding-Standards/pull/624): ProperEscapingFunction: privatize $escaping_functions property along with various improvements.

Reviewer: This PR added a new check (and error code) to the sniff, so should probably have a separate entry in "Added" and explicitly mention the new functionality (checking for use of `esc_attr()` when escaping something which is not an HTML attribute will now be flagged).

The entry here in "Changed" should probably mention that the sniff now takes more escaping functions into account.

And... I think a separate entry in "Fixed" is also warranted as the PR also fixes various bugs, like where the escaping when HTML attributes used slash escaped quotes in the PHP text string would not correctly be handled/recognized or where the strings using double quotes were not handled correctly etc etc.

- [#635](https://github.com/Automattic/VIP-Coding-Standards/pull/635): Ruleset: remove WordPress.WP.GlobalVariablesOverride.Prohibited severity declaration.

> - [#638](https://github.com/Automattic/VIP-Coding-Standards/pull/638): IncludingFile: allow custom keywords to reduce false positives.

Author: Noted and since this is new functionality it would belong in the "Added" section.

- [#644](https://github.com/Automattic/VIP-Coding-Standards/pull/644): RestrictedFunctions: remove wp_cache_get_multi group.
- [#645](https://github.com/Automattic/VIP-Coding-Standards/pull/645): Ruleset: silence WordPress.WP.AlternativeFunctions.file_system_read_readfile.
- [#646](https://github.com/Automattic/VIP-Coding-Standards/pull/646): Ruleset: silence WordPress.WP.AlternativeFunctions.file_system_read_fclose.
- [#647](https://github.com/Automattic/VIP-Coding-Standards/pull/647): RestrictedFunctions: remove get_super_admins group.
- [#649](https://github.com/Automattic/VIP-Coding-Standards/pull/649): RestrictedFunctions: downgrade switch_to_blog() to warning and change messaging.
- [#652](https://github.com/Automattic/VIP-Coding-Standards/pull/652): RestrictedFunctions/RestrictedVariables: remove usermeta related errors.

## Fixed
- [#444](https://github.com/Automattic/VIP-Coding-Standards/pull/444): ConstantString: account for namespace prefixing.
- [#581](https://github.com/Automattic/VIP-Coding-Standards/pull/581): AlwaysReturnInFilter: fix runtime failure on abstract methods.

> - [#586](https://github.com/Automattic/VIP-Coding-Standards/pull/586): IncludingNonPHPFile: various bug fixes.

Author: Since there are many (6) included in this PR, listing all of them seems like it would make for a long-winded entry. However, I'm not opposed to calling out a few of them. Which bug fixes in that PR would you prioritize calling out? I'm leaning towards the below:
- case sensitivity of file extension
- allow for "phar" file extension
- extensions in interpolated strings

> - [#586](https://github.com/Automattic/VIP-Coding-Standards/pull/586): IncludingNonPHPFile: various bug fixes.

Reviewer: The "phar" should probably be in the "Changed" entry. I agree on the other for the "Fixed" entry.

> - [#586](https://github.com/Automattic/VIP-Coding-Standards/pull/586): IncludingNonPHPFile: various bug fixes.

Author: I'm leaning towards the phar file as in "Added" since it's technically a new feature, no?

- [#587](https://github.com/Automattic/VIP-Coding-Standards/pull/587): LowExpiryCacheTime: allow arithmetic operators, simple floats, numerical strings, zeroes and parantheses in calculations, and FQN time constants.
- [#592](https://github.com/Automattic/VIP-Coding-Standards/pull/592): DynamicCalls: ignore comments, allow double quotes and remove potential memory leak.

> - [#595](https://github.com/Automattic/VIP-Coding-Standards/pull/595): Underscorejs: no error shown during escaped variable.

Reviewer: Looks like the original entry got _moved_,, while it should have remained in "Fixed". A second entry was needed for the added functionality.

> - [#654](https://github.com/Automattic/VIP-Coding-Standards/pull/654): LowExpiryCacheTime: improve handling of parse errors in evaluated code.

Reviewer: This doesn't need an entry as it is a bug fix cleaning up after #587. I.e. the bug was never in a release.

## General Housekeeping
- [#582](https://github.com/Automattic/VIP-Coding-Standards/pull/582): CI: re-try composer install on failure.

> - [#583](https://github.com/Automattic/VIP-Coding-Standards/pull/583): Composer: require DirectDealer plugin.

Author: I can definitely move it to the "Added" section but I do feel like that recommendation is too wordy for a changelog bullet point entry. The recommendation would be better lived above the changelog as a disclaimer.

> - [#583](https://github.com/Automattic/VIP-Coding-Standards/pull/583): Composer: require DirectDealer plugin.

Reviewer: Agreed.

- [#585](https://github.com/Automattic/VIP-Coding-Standards/pull/585): AlwaysReturnInFilter: adjust test expectations for upstream tokenizer bugfix.
- [#596](https://github.com/Automattic/VIP-Coding-Standards/pull/596): DynamicCalls: use "disallowed" wording over "blacklisted".
- [#599](https://github.com/Automattic/VIP-Coding-Standards/pull/599): CI: add build against PHP 8.
- [#606](https://github.com/Automattic/VIP-Coding-Standards/pull/606): Ruleset: remove redundant rule ref.
- [#607](https://github.com/Automattic/VIP-Coding-Standards/pull/607): Ruleset: remove redundant rule ref.
- [#608](https://github.com/Automattic/VIP-Coding-Standards/pull/608): Ruleset: remove duplicate rule ref.
- [#611](https://github.com/Automattic/VIP-Coding-Standards/pull/611): Ruleset: remove redundant severity declaration.
- [#617](https://github.com/Automattic/VIP-Coding-Standards/pull/617): Ruleset: remove redundant severity declaration.

> - [#618](https://github.com/Automattic/VIP-Coding-Standards/pull/618): Remove Batcache references from error messages.

Reviewer: The old entry is still in the "Housekeeping" section. Either it should be removed, or the reference to the change in the messaging should be removed from the entry in "Changed" as it is now mentioned twice.

- [#619](https://github.com/Automattic/VIP-Coding-Standards/pull/619): Docs: Update links to wpvip.com.

> - [#630](https://github.com/Automattic/VIP-Coding-Standards/pull/630): CI: fix test failure for VIPCS VariableAnalysis sniff.

Reviewer: The text of the entry (now in "Changed") still reads "CI: ...". This wasn't a CI change, but a change to the deprecated sniff.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

> # [2.3.0] - 2021-04-19

Reviewer 2: ```suggestion
## [2.3.0] - 2021-04-19
```

Props: jrfnl, rebeccahum, kevinfodness, GaryJones.

** There is a minor breaking change in the ProperEscapingFunction sniff from PR #624. The `escaping_function` property can no longer be overruled via custom rulesets. Please remove any usages of the property in custom rulesets.

> ** Composer now requires the DirectDealer plugin per #583. Note: If you either include it in the "require-dev" of your composer.json, use another Composer PHPCS plugin, or run bash commands to register PHPCS standards, please remove it from those sources to prevent interferences or version constraint conflicts.

Reviewer 2: ```suggestion
** Composer now requires the [phpcodesniffer-composer-installer](https://github.com/Dealerdirect/phpcodesniffer-composer-installer) plugin per #583. Note: If you either include it in the "require-dev" of your `composer.json`, use another Composer PHPCS plugin, or run bash commands to register PHPCS standards, please remove it from those sources to prevent interferences or version constraint conflicts.
```

> ## Added

Reviewer 2: ```suggestion
### Added
```

- [#581](https://github.com/Automattic/VIP-Coding-Standards/pull/581): AlwaysReturnInFilter: flag abstract methods for manual inspection.

> - [#583](https://github.com/Automattic/VIP-Coding-Standards/pull/583): Composer: require DirectDealer plugin.

Reviewer 2: ```suggestion
- [#583](https://github.com/Automattic/VIP-Coding-Standards/pull/583): Composer: require phpcs-composer-installer plugin.
```

- [#589](https://github.com/Automattic/VIP-Coding-Standards/pull/589): WPQueryParams: flags 'exclude' array key.

> - [#595](https://github.com/Automattic/VIP-Coding-Standards/pull/595): Underscorejs: fixed false positive for when a variable is `_.escape()`-ed.

Reviewer 2: ```suggestion
- [#595](https://github.com/Automattic/VIP-Coding-Standards/pull/595): Underscore.js: fixed false positive for when a variable is `_.escape()`-ed.
```

Reviewer: Not sure if this suggestion is correct as the sniff _name_ doesn't contain the `.` and these references at the start of a changelog entry normally refer to the sniff name. What do you think @GaryJones ?

Author: Agreed with @jrfnl.

- [#624](https://github.com/Automattic/VIP-Coding-Standards/pull/624): ProperEscapingFunction: account for additional escaping functions and check for `esc_attr()` usage in non-HTML attributes.

> - [#638](https://github.com/Automattic/VIP-Coding-Standards/pull/638): IncludingFile: new public property `$allowedKeywords` for custom partial keywords in constants to reduce false positives.

Reviewer 2: ```suggestion
- [#638](https://github.com/Automattic/VIP-Coding-Standards/pull/638): IncludingFile: new public property `$allowedKeywords` for allowing custom partial keywords in constants to reduce false positives.
```

> ## Changed

Reviewer 2: ```suggestion
### Changed
```

- [#586](https://github.com/Automattic/VIP-Coding-Standards/pull/586): IncludingNonPHPFile: various performance improvements.
- [#587](https://github.com/Automattic/VIP-Coding-Standards/pull/587): LowExpiryCacheTime: new warning added for manual inspection along with various improvements.
- [#592](https://github.com/Automattic/VIP-Coding-Standards/pull/592): DynamicCalls: various improvements.

> - [#595](https://github.com/Automattic/VIP-Coding-Standards/pull/595): Underscorejs: various improvements.

Reviewer 2: ```suggestion
- [#595](https://github.com/Automattic/VIP-Coding-Standards/pull/595): Underscore.js: various improvements.
```

Reviewer: Dito.

- [#630](https://github.com/Automattic/VIP-Coding-Standards/pull/630): CI: fix incompatibility for VariableAnalysis standard with previously deprecated native VIPCS VariableAnalysis sniff.
- [#639](https://github.com/Automattic/VIP-Coding-Standards/pull/639): RestrictedFunctions: remove site_option group.
- [#644](https://github.com/Automattic/VIP-Coding-Standards/pull/644): RestrictedFunctions: remove wp_cache_get_multi group.
- [#645](https://github.com/Automattic/VIP-Coding-Standards/pull/645): Ruleset: silence WordPress.WP.AlternativeFunctions.file_system_read_readfile.
- [#646](https://github.com/Automattic/VIP-Coding-Standards/pull/646): Ruleset: silence WordPress.WP.AlternativeFunctions.file_system_read_fclose.
- [#647](https://github.com/Automattic/VIP-Coding-Standards/pull/647): RestrictedFunctions: remove get_super_admins group.
- [#649](https://github.com/Automattic/VIP-Coding-Standards/pull/649): RestrictedFunctions: downgrade switch_to_blog() to warning and change messaging.
- [#652](https://github.com/Automattic/VIP-Coding-Standards/pull/652): RestrictedFunctions/RestrictedVariables: remove usermeta related errors.

> ## Fixed

Reviewer 2: ```suggestion
### Fixed
```

- [#444](https://github.com/Automattic/VIP-Coding-Standards/pull/444): ConstantString: only error when a plain constant is passed as constant name parameter.
- [#581](https://github.com/Automattic/VIP-Coding-Standards/pull/581): AlwaysReturnInFilter: fix runtime failure on abstract methods.
- [#584](https://github.com/Automattic/VIP-Coding-Standards/pull/584): Performance: more selective sniffing for efficiency.
- [#586](https://github.com/Automattic/VIP-Coding-Standards/pull/586): IncludingNonPHPFile: various bug fixes such as recognition of .phar extensions, interpolated strings, and case insensitivity used in file extensions. 
- [#587](https://github.com/Automattic/VIP-Coding-Standards/pull/587): LowExpiryCacheTime: allow arithmetic operators, simple floats, numerical strings, zeroes and parentheses in calculations, and FQN time constants.
- [#592](https://github.com/Automattic/VIP-Coding-Standards/pull/592): DynamicCalls: ignore comments, allow double quotes and remove potential memory leak.
- [#624](https://github.com/Automattic/VIP-Coding-Standards/pull/624): ProperEscapingFunction: slash escaped quotes and non-quoted strings in HTML attributes are now parsed as expected.

> ## Removed

Reviewer 2: ```suggestion
### Removed
```

- [#624](https://github.com/Automattic/VIP-Coding-Standards/pull/624): ProperEscapingFunction: remove `$escaping_functions` public property.

> ## Housekeeping

Reviewer 2: ```suggestion
### Maintenance
```
